robotframework-ls/src/robotframework_ls/impl/flow_explorer_model_builder.py

```diff
@@ -1,9 +1,53 @@
+from contextlib import contextmanager
 from robotframework_ls.impl.protocols import ICompletionContext, IKeywordFound, INode
 from typing import List, Any, Dict, Tuple
 import os
 from robocorp_ls_core.basic import isinstance_name
 
 
+class _KeywordRecursionStack:
+    _stack: Dict
+
+    def __init__(self) -> None:
+        self._stack = {}
+
+    def has(self, keyword_name: str) -> bool:
+        return keyword_name in self._stack
+
+    def push(self, keyword_name: str) -> None:
+        self._stack.update({keyword_name: None})
+
+    def pop(self, keyword_name: str) -> None:
+        self._stack.pop(keyword_name)
+
+    @contextmanager
+    def scoped(self, keyword_name):
+        self.push(keyword_name)
+        yield
+        self.pop(keyword_name)
+
+
+class _UserKeywordCollector:
+    _user_keywords: List[Any]
+    _name_collection: Dict
+
+    def __init__(self) -> None:
+        self._user_keywords = []
+        self._name_collection = {}
+
+    @property
+    def keywords(self) -> List[Any]:
+        return self._user_keywords
+
+    def has(self, keyword_name) -> bool:
+        return keyword_name in self._name_collection
+
+    def append(self, keyword: Any) -> None:
+        if "name" in keyword:
+            self._user_keywords.append(keyword)
+            self._name_collection.update({keyword["name"]: None})
```

Comment:
After changing `_name_collection` for a `set` use `self._name_collection.add(keyword['name`]).